### PR TITLE
lib_pathbuffer.c:Use atomic instead of locks

### PR DESCRIPTION
**Pull request by crafcat7**
## Summary

Avoid spinlock/mutex by using atomic

Use atomic_cmpxchg to ensure that in multithreaded situations, if someone releases the buffer, it can be applied for in time. And use atomic_ulong to save free_bitmap

## Impact

**Is a new feature added?:** NO
**Impact on build:** NO
**Impact on hardware:** NO, this change does not specifically target any particular hardware architectures or boards.
**Impact on documentation:** NO,This patch does not introduce any new features
**Impact on compatibility:** This implementation aims to be backward compatible. No existing functionality is expected to be broken.

## Testing

Build Host(s): Linux x86
Target(s): sim/nsh

Passed the file system related syscall test

### CODE REVIEW

File: libs/libc/misc/lib_pathbuffer.c [hunk 1]
```diff
@@ -82,21 +80,24 @@ static struct pathbuffer_s g_pathbuffer =
...
-  irqstate_t flags;
   int index;
+  uint32_t free_bitmap;
 
-  /* Try to find a free buffer */
-
-  flags = spin_lock_irqsave(&g_pathbuffer.lock);
-  index = ffsl(g_pathbuffer.free_bitmap) - 1;
-  if (index >= 0 && index < CONFIG_LIBC_PATHBUFFER_MAX)
+  do
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_pathbuffer.c:86]: for (; ; )

File: libs/libc/misc/lib_pathbuffer.c [hunk 2]
```diff
@@ -82,21 +80,24 @@ static struct pathbuffer_s g_pathbuffer =
...
     {
-      g_pathbuffer.free_bitmap &= ~(1u << index);
-      spin_unlock_irqrestore(&g_pathbuffer.lock, flags);
-      return g_pathbuffer.buffer[index];
+      free_bitmap = atomic_read(&g_pathbuffer.free_bitmap);
+      index = ffsl(free_bitmap) - 1;
+      if (index >= 0 && index < CONFIG_LIBC_PATHBUFFER_MAX)
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_pathbuffer.c:90]:
```
if (index < 0)
  {
     break;
  }
```

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



Yes, this PR appears to meet the NuttX requirements, although the testing section could be improved.

While the PR provides information on the host and target, it lacks specific "before" and "after" testing logs.  Simply stating "Passed the file system related syscall test" is insufficient.  Concrete examples of commands run and their output, both before and after the change, are crucial for reviewers to assess the impact and correctness of the patch.  Even if the output is identical, showing the commands and output demonstrates that the tests were actually run.

File: libs/libc/misc/lib_pathbuffer.c (hunk 1 above)
Comment (crafcat7) [reply libs/libc/misc/lib_pathbuffer.c]: Done

File: libs/libc/misc/lib_pathbuffer.c (hunk 2 above)
Comment (crafcat7) [reply libs/libc/misc/lib_pathbuffer.c]: Done

File: libs/libc/misc/lib_pathbuffer.c [hunk 3]
```diff
@@ -82,22 +80,26 @@ static struct pathbuffer_s g_pathbuffer =
 
 FAR char *lib_get_pathbuffer(void)
 {
-  irqstate_t flags;
   int index;
+  int32_t free_bitmap;
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_pathbuffer.c:84]: merge to line 88

File: libs/libc/misc/lib_pathbuffer.c [hunk 4]
```diff
@@ -82,22 +80,26 @@ static struct pathbuffer_s g_pathbuffer =
 
 FAR char *lib_get_pathbuffer(void)
 {
-  irqstate_t flags;
   int index;
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_pathbuffer.c:83]: merge to line 89

File: libs/libc/misc/lib_pathbuffer.c [hunk 5]
```diff
@@ -82,22 +80,26 @@ static struct pathbuffer_s g_pathbuffer =
...
-      spin_unlock_irqrestore(&g_pathbuffer.lock, flags);
-      return g_pathbuffer.buffer[index];
+      free_bitmap = atomic_read(&g_pathbuffer.free_bitmap);
+      index = ffsl(free_bitmap) - 1;
+      if (index < 0 || index >= CONFIG_LIBC_PATHBUFFER_MAX)
+        {
+          break;
+        }
+
+      int32_t update = free_bitmap & ~(1u << index);
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_pathbuffer.c:95]: move the declaration to line 94

**Review by anchao — APPROVED** [review]

File: libs/libc/misc/lib_pathbuffer.c (hunk 5 above)
Comment (crafcat7) [reply libs/libc/misc/lib_pathbuffer.c]: Done

File: libs/libc/misc/lib_pathbuffer.c (hunk 4 above)
Comment (crafcat7) [reply libs/libc/misc/lib_pathbuffer.c]: Done

File: libs/libc/misc/lib_pathbuffer.c [hunk 6]
```diff
@@ -82,22 +80,24 @@ static struct pathbuffer_s g_pathbuffer =
...
-      spin_unlock_irqrestore(&g_pathbuffer.lock, flags);
-      return g_pathbuffer.buffer[index];
+      int32_t free_bitmap = atomic_read(&g_pathbuffer.free_bitmap);
+      int index = ffsl(free_bitmap) - 1;
+      if (index < 0 || index >= CONFIG_LIBC_PATHBUFFER_MAX)
+        {
+          break;
+        }
+
+      int32_t update;
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_pathbuffer.c:92]: move to the beginning of loop

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**